### ISSUE TITLE
close dialog at the same time close the dependency

### ISSUE THREAD

**Issue by yangxuy** [jonataslaw/getx#1969]
when i close a dialog at the same time the page's dependency has closed

**Comment by FlorianHeuse** [comment]
It is hard to follow without code. Thanks for your understanding.

**Comment by fooshen-chong** [comment]
i face this problem as well. Attached with a minimal example.

`import 'package:flutter/material.dart';
import 'package:get/get.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return GetMaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(),
    );
  }
}

class SubController extends GetxController {}

class Controller extends GetxController {
  Future<void> load() async {
    Get.dialog<dynamic>(
        Center(
          child: SizedBox(width: 120, height: 120, child: Container()),
        ),
        barrierDismissible: false);
    Get.put(SubController());
    Future.delayed(Duration(seconds: 3));
    Get.back<dynamic>();

    Get.bottomSheet<dynamic>(Container());
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final Controller controller = Get.put(Controller());
    return Scaffold(
        body: Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          TextButton(
              onPressed: () {
                controller.load();
              },
              child: Text('LOAD')),
        ],
      ),
    ));
  }
}
`

Notice that the debug message : 
[GETX] OPEN DIALOG 89612746
[GETX] Instance "SubController" has been created
[GETX] Instance "SubController" has been initialized
[GETX] CLOSE DIALOG 89612746
[GETX] OPEN BOTTOMSHEET 798650144
[GETX] "SubController" onDelete() called
[GETX] "SubController" deleted from memory

after the close dialog called, to close the overlay dialog, the inited subcontroller get deleted as well

**Comment by FlorianHeuse** [comment]
> i face this problem as well. Attached with a minimal example.
> 
> ```
> import 'package:flutter/material.dart';
> import 'package:get/get.dart';
> 
> void main() {
>   runApp(MyApp());
> }
> 
> class MyApp extends StatelessWidget {
>   // This widget is the root of your application.
>   @override
>   Widget build(BuildContext context) {
>     return GetMaterialApp(
>       title: 'Flutter Demo',
>       home: MyHomePage(),
>     );
>   }
> }
> 
> class SubController extends GetxController {}
> 
> class Controller extends GetxController {
>   Future<void> load() async {
>     Get.dialog<dynamic>(
>         Center(
>           child: SizedBox(width: 120, height: 120, child: Container()),
>         ),
>         barrierDismissible: false);
>     Get.put(SubController());
>     Future.delayed(Duration(seconds: 3));
>     Get.back<dynamic>();
>     Get.bottomSheet<dynamic>(Container());
>   }
> }
> 
> class MyHomePage extends StatelessWidget {
>   @override
>   Widget build(BuildContext context) {
>     final Controller controller = Get.put(Controller());
>     return Scaffold(
>         body: Center(
>       child: Column(
>         mainAxisAlignment: MainAxisAlignment.center,
>         children: [
>           TextButton(
>               onPressed: () {
>                 controller.load();
>               },
>               child: Text('LOAD')),
>         ],
>       ),
>     ));
>   }
> }
> ```
> 
> Notice that the debug message : [GETX] OPEN DIALOG 89612746 [GETX] Instance "SubController" has been created [GETX] Instance "SubController" has been initialized [GETX] CLOSE DIALOG 89612746 [GETX] OPEN BOTTOMSHEET 798650144 [GETX] "SubController" onDelete() called [GETX] "SubController" deleted from memory
> 
> after the close dialog called, to close the overlay dialog, the inited subcontroller get deleted as well

You run into a timing problem. Get.dialog returns a future (Future<T?> dialog<T>). You could use two solutions (1) 
```
    await Get.dialog<dynamic>(
        Center(
          child: SizedBox(width: 120, height: 120, child: Container()),
        ),
        barrierDismissible: false);
```

Or solution (2) - make your dependency permanent:
```
Get.put(SubController(), permanent: true);
```

**Comment by fooshen-chong** [comment]
''''    
await Get.dialog<dynamic>(
        Center(
          child: SizedBox(width: 120, height: 120, child: Container()),
        ),
        barrierDismissible: false);

''''

Thanks for the prompt reply.
if add await to Get.dialog it will not trigger Get.back as it is always awaiting , the dialog will not be close at all, or my understanding is wrong?

''''
Get.put(SubController(), permanent: true);

''''

i think this is not a good suggestion if subcontroller contains heavy processes inside.

**Comment by fooshen-chong** [comment]
@jonataslaw  sorry for bumping this.
